File: backend/core/use_cases/snapshots.py
from datetime import datetime
from uuid import UUID

from core.domain.models import Snapshot
from core.ports.repositories import AccountRepository, SnapshotRepository


class SnapshotService:
    def __init__(self, repo: SnapshotRepository, account_repo: AccountRepository) -> None:
        self.repo = repo
        self.account_repo = account_repo

    def record(
        self,
        account_id: UUID,
        deposit: float,
        unrealized_return: float = 0.0,
        recorded_at: datetime | None = None,
    ) -> Snapshot:
        if not self.account_repo.get_by_id(account_id):
            raise ValueError(f"Account {account_id} not found")
        kwargs: dict = {"account_id": account_id, "deposit": deposit, "unrealized_return": unrealized_return}
        if recorded_at is not None:
            kwargs["recorded_at"] = recorded_at
        return self.repo.save(Snapshot(**kwargs))

    def get_history(self, account_id: UUID) -> list[Snapshot]:
        if not self.account_repo.get_by_id(account_id):
            raise ValueError(f"Account {account_id} not found")
        return self.repo.get_for_account(account_id)

    def update(
        self,
        account_id: UUID,
        snapshot_id: UUID,
        deposit: float,
        unrealized_return: float,
        recorded_at: datetime,
    ) -> Snapshot:
        if not self.account_repo.get_by_id(account_id):
            raise ValueError(f"Account {account_id} not found")
        snapshot = Snapshot(
            id=snapshot_id,
            account_id=account_id,
            deposit=deposit,
            unrealized_return=unrealized_return,
            recorded_at=recorded_at,
        )
        return self.repo.update(snapshot)

    def delete(self, account_id: UUID, snapshot_id: UUID) -> None:
        if not self.account_repo.get_by_id(account_id):
            raise ValueError(f"Account {account_id} not found")
        self.repo.delete(account_id, snapshot_id)

    def get_latest_for_all(self) -> dict[UUID, Snapshot]:
        return self.repo.get_latest_for_all()

File: frontend/lib/repositories/snapshots.ts
import { getDb, type StoredSnapshot } from "../db"
import type { Snapshot, UpdateSnapshotInput } from "../types"

function toSnapshot(s: StoredSnapshot): Snapshot {
  return { ...s, total: s.deposit + s.unrealized_return }
}

export const snapshotsRepo = {
  async getAll(): Promise<Snapshot[]> {
    const db = await getDb()
    return (await db.getAll("snapshots")).map(toSnapshot)
  },

  async getForAccount(accountId: string): Promise<Snapshot[]> {
    const db = await getDb()
    const items = await db.getAllFromIndex("snapshots", "by_account", accountId)
    return items.map(toSnapshot).sort((a, b) => a.recorded_at.localeCompare(b.recorded_at))
  },

  async getLatestForAll(): Promise<Record<string, Snapshot>> {
    const all = await this.getAll()
    const result: Record<string, Snapshot> = {}
    for (const snap of all) {
      const existing = result[snap.account_id]
      if (!existing || snap.recorded_at > existing.recorded_at) {
        result[snap.account_id] = snap
      }
    }
    return result
  },

  async getEarliestDate(): Promise<Date | null> {
    const db = await getDb()
    const all = await db.getAll("snapshots")
    if (all.length === 0) return null
    const earliest = all.reduce(
      (min, s) => (s.recorded_at < min ? s.recorded_at : min),
      all[0].recorded_at,
    )
    return new Date(earliest)
  },

  async save(snapshot: Snapshot): Promise<Snapshot> {
    const db = await getDb()
    const { total: _, ...stored } = snapshot
    await db.put("snapshots", stored)
    return snapshot
  },

  async update(snapshotId: string, data: UpdateSnapshotInput): Promise<Snapshot> {
    const db = await getDb()
    const existing = await db.get("snapshots", snapshotId)
    if (!existing) throw new Error(`Snapshot ${snapshotId} not found`)
    const updated = { ...existing, ...data }
    await db.put("snapshots", updated)
    return toSnapshot(updated)
  },

  async delete(snapshotId: string): Promise<void> {
    const db = await getDb()
    await db.delete("snapshots", snapshotId)
  },
}
